### ISSUE TITLE
Qwen3-32B: intermittent corrupted output at top_p=1.0 with on-device sampling

### ISSUE THREAD

**Issue by ksriv001** [tenstorrent/tt-metal#50649]
# Qwen3-32B: intermittent corrupted output at `top_p=1.0` with on-device sampling

## Summary

`Qwen/Qwen3-32B` intermittently produces corrupted output when a request runs at **`top_p=1.0`** with on-device sampling (`sample_on_device_mode=all`). We see it two ways:

- garbled/junk tokens in the generated text, and
- the model emitting **EOS right after a think block closes** — in a tools + thinking request this shows up as a **dropped tool call** (`finish_reason: stop`, empty content, no `<tool_call>`).

The one clear signal we have is that it's a function of `top_p`: at `top_p=1.0` it happens 17–35% of the time, and it **disappears once `top_p` is clamped** (`top_p ≤ 0.95` → 0 failures across our runs). The identical model on another host (Nebius via OpenRouter) shows 0 failures at `top_p=1.0`, so this doesn't look like ordinary sampling variance.

We haven't dug into the device sampling path ourselves — filing this so someone who knows that code can take a look. Repro steps and the data we have are below.

Related: [tt-inference-server#4386](https://github.com/tenstorrent/tt-inference-server/issues/4386) (Issue 4) and [tt-cloud-console#944](https://github.com/tenstorrent/tt-cloud-console/pull/944).

## Environment

- Model: `Qwen/Qwen3-32B`
- On-device sampling: `sample_on_device_mode=all`
- Flags: `--enable-auto-tool-choice --tool-call-parser hermes --reasoning-parser qwen3`
- Transport: streaming SSE
- Reproduced with **tt-metal = stable**, tt-inference-server = `lzhang/openrouter-issue-4-stable` (tests on top of stable), vLLM = dev.

## Where this shows up (tt-shield)

The drop reproduces under release-dispatch shield runs on **tt-metal=stable** (tt-inference-server = `lzhang/openrouter-issue-4-stable`, which carries the tests), and it **persists across the vLLM streaming-parser fix** — which isolates it from the reasoning/tool streaming parser:

- vLLM=dev (no parser fix): issues 1, 2, and 4 all appear — [run 29852446182](https://github.com/tenstorrent/tt-shield/actions/runs/29852446182)
- vLLM=`lzhang/qwen3-stream-parser` (parser fix for issues 1 & 2): issues 1 & 2 clear, but **issue 4 still appears** — [run 29855159423](https://github.com/tenstorrent/tt-shield/actions/runs/29855159423)

(Issues 1 & 2 are the streaming reasoning-parser leak, fixed by the vLLM parser branch. This ticket is only about issue 4 — the intermittent `top_p=1.0` drop — which the parser fix does **not** resolve.)

## How to reproduce

Send the same chat-completions request `N=20+` times at `top_p=1.0`, then repeat at `top_p=0.95`. The payload below is the tools + thinking case (where the symptom is easiest to score — a dropped tool call), but the corruption also shows up in the generated text of plain `top_p=1.0` requests.

```bash
curl -s -N "$ENDPOINT/v1/chat/completions" \
  -H "Authorization: Bearer $TOKEN" \
  -H "Content-Type: application/json" \
  -d '{
    "model": "Qwen/Qwen3-32B",
    "stream": true,
    "stream_options": {"include_usage": true},
    "messages": [
      {"role": "system", "content": "You must think before every response. First write at least one sentence of reasoning, then give a concise final answer."},
      {"role": "user", "content": "What is the weather like in Boston, MA in fahrenheit? Use the get_weather tool."}
    ],
    "max_tokens": 1024,
    "temperature": 1, "top_p": 1,
    "tools": [{"type": "function", "function": {"name": "get_weather", "description": "Get the current weather in a given location", "parameters": {"type": "object", "properties": {"location": {"type": "string"}, "unit": {"type": "string", "enum": ["celsius", "fahrenheit"]}}, "required": ["location", "unit"]}}}],
    "tool_choice": "auto",
    "chat_template_kwargs": {"enable_thinking": true}
  }'
```

Count runs where `finish_reason == "stop"` with no `tool_calls` delta (dropped tool call). Then flip `"top_p": 1` → `"top_p": 0.95` and repeat.

## What we observed (`Qwen/Qwen3-32B`, identical payload)

| Config | tool call emitted | failure rate |
|---|---|---|
| `top_p=1.0` | 11–16/20 | **17–35%** |
| `top_p=0.95` | 20/20 | 0% |
| Nebius via OpenRouter (`qwen/qwen3-32b`), `top_p=1.0` | 20/20 | 0% |

Dose–response over `top_p` (N=30 each) — failures fall off as `top_p` is lowered:

| top_p | 1.0 | 0.99 | 0.98 | 0.96 | 0.95 |
|---|---|---|---|---|---|
| fail rate | 17% | 0% | 3% | 0% | 0% |

Two failure sub-modes at `top_p=1.0`: **dominant (~8/9)** the tool call is dropped entirely (`finish:stop`, empty content, no `<tool_call>`); **minority (~1/9)** the tool-call XML is emitted but stranded in `reasoning_content` (think block never closes → hermes never parses it).

## What we'd like

Someone familiar with the on-device sampling path to reproduce this and look into why `top_p=1.0` produces corrupted tokens where `top_p ≤ 0.95` doesn't. Happy to provide an endpoint, JWT, and a scripted N-run harness if useful.

**Comment by tchedaTT** [comment]
Running with tt-metal `main` and vllm `dev`, I am not seeing this issue - failures are happening due to dropping either </think> or <tool_call>. top_1=1.0 failed 15/200, 12x missing </think>, 3x missing <tool_call>. top_p=0.95 failed 9/200, 9x misssing </think>. Switching to lzhang/qwen3-stream-parser to isolate the issue

**Comment by tchedaTT** [comment]
Running with tt-metal main and vllm lzhang/openrouter-issue-4-stable, failures with missing </think> occur both with top_p 1.0 and 0.95, both 6/100, no missing <tool_call>
Running with tt-metal stable and vllm lzhang/openrouter-issue-4-stable all failures are missing <think>
Streaming: 1.0 5/100, 0.95 1/100. Non-streaming 1.0 8/100, 0.95 4/100.
Running stable+lzhang/openrouter-issue-4-stable with host sampling there are no failures, confirming the issue lies with the sampling module.
I'm investigating further, however, the failure rate I'm seeing is much lower than reported here. Could you share your commands for reproducing with just vllm, no inference-server?

**Comment by tchedaTT** [comment]
Correction: there was 1 failure in 400 runs when running host sampling, it was a streaming top_p=1.0 run

**Comment by ksriv001** [comment]
@tchedaTT Thank you for looking into this. Would there be any difference in the outputs if we are doing sampling on device rather than on host? We are running the model on WH Galaxy for which device side sampling is supported.

**Comment by tchedaTT** [comment]
Sampling on device and on host should be equivalent for all supported cases (and all unsupported should fall back).
I can see that even with host sampling failures occur - got 5 in 771 requests, 0.65% failure rate. I looked at logprobs in these cases - tool call stayed in reasoning, and <tool call> genuinely had the highest logits, higher than </think> by some margin. In the successful cases, </think> was dominant by a wide margin.

I can't easily inspect the logits in the device sampling case.
Now checking the hypothesis that there is some trace-related race or corruption.

**Comment by tchedaTT** [comment]
I tested with device sampling enabled, but tracing disabled. Failed requests 2/473, roughly matching the failure ratio of host sampling.
To summarize:
Explicit host sampling 1/400
Logprobs (fallback to host sampling) 5/722
In total host sampling - 6/1172=~0.5%

Device sampling + trace disabled 2/473=0.42%


Device sampling, all with lzhang/qwen3-stream-parser:
dev tt-metal 28/400 ~7%
stable tt-metal 18/400 ~4.5%
breaking down further
```
+--------+---------------+-----------------+-----------+
|        | Streaming     | Non-streaming   | Total     |
+--------+---------------+-----------------+-----------+
| Dev    | 12/200 (6.0%) | 16/200 (8.0%)   | 28/400    |
| Stable |  6/200 (3.0%) | 12/200 (6.0%)   | 18/400    |
+--------+---------------+-----------------+-----------+
| Total  | 18/400 (4.5%) | 28/400 (7.0%)   | 46/800    |
+--------+---------------+-----------------+-----------+
```
```
+--------+----------------+----------------+-----------+
|        | top_p=1.0      | top_p=0.95     | Total     |
+--------+----------------+----------------+-----------+
| Dev    | 18/200 (9.0%)  | 10/200 (5.0%)  | 28/400    |
| Stable | 13/200 (6.5%)  |  5/200 (2.5%)  | 18/400    |
+--------+----------------+----------------+-----------+
| Total  | 31/400 (7.75%) | 15/400 (3.75%) | 46/800    |
+--------+----------------+----------------+-----------+
```

My suspicion now is that there is either a data race or a corruption due to tracing which boosts the probability of <tool> vs </think>.

**Comment by tchedaTT** [comment]
This might share the same root cause as https://github.com/tenstorrent/tt-metal/issues/37397

**Comment by tchedaTT** [comment]
I just found that upstream contains a workaround for this issue - apparently Qwen will sometimes just forget to close reasoning. If we cherrypick this, it should solve this problem https://github.com/vllm-project/vllm/pull/35687

**Comment by tchedaTT** [comment]
In my testing this seems to solve, or at least mask, the issue. Merged the cherrypick onto dev. Please check if you can reproduce

**Comment by ksriv001** [comment]
@lukezTT Can you please mention the link to your CI run with these changes here? Would be good to have all of it here for context :)

**Comment by lukezTT** [comment]
This run has [Cherrypick Qwen3ReasoningParser fix](https://github.com/tenstorrent/vllm/commit/9bd099cbda4872bba1abdb43eea859f88c05330c) + run on stable with top_p=1, which we expect to pass issues 1/2 but we are still seeing issue 1 fail: https://github.com/tenstorrent/tt-shield/actions/runs/30012231413/job/89241235726

To isolate for top_p interference, this run has the same [Cherrypick Qwen3ReasoningParser fix](https://github.com/tenstorrent/vllm/commit/9bd099cbda4872bba1abdb43eea859f88c05330c) + run on stable **with top_p=0.95**, but we still see issue 1 fail: https://github.com/tenstorrent/tt-shield/actions/runs/30039972846/job/89317296196

This run has @tchedaTT fixes to vllm, on top of [Cherrypick Qwen3ReasoningParser fix](https://github.com/tenstorrent/vllm/commit/9bd099cbda4872bba1abdb43eea859f88c05330c) + run on stable with top_p=1: https://github.com/tenstorrent/tt-shield/actions/runs/30119798946

**Comment by tchedaTT** [comment]
Note: the failing test https://github.com/tenstorrent/tt-shield/actions/runs/30119798946 only asks the model to repeat something in the style of the provided json and uses response_format json_object, which uses structured outputs and switches to host sampling to do it. It does not actually do tool calling.

**Comment by tchedaTT** [comment]
Potentially related: https://github.com/tenstorrent/vllm-tt-plugin/pull/10/changes

**Comment by feslamiTT** [comment]
Could someone summarize the situation as it stands?  Are we saying that the bug for this issue (intermittent corrupted output) was not related to tool calling?  Or that it was, and it has been fixed, and now the test suite is exposing a _different_ issue?  Does that suite surface all the failures, or were we never testing the 2nd problem because the first one was cancelling the entire job?

**Comment by mtairum** [comment]
@ifabijanic-tt to update. Related to https://github.com/tenstorrent/tt-metal/issues/49785

**Comment by ifabijanic-tt** [comment]
The issue was being tested with latest build as of the day of writing this. At the time build info:
tt-metal: 5b8d933de02
vLLM: 0.26.0
tt-vllm-plugin: d7a6008

TLDR:
1,028 requests, 0 failures on the issue's own criterion with no dropped <tool_call>, no unclosed <think>, across six configurations.

| Arm (device sampling, trace on) | N | no </think> | no <tool_call> |
|---------------------------------|---|------------|----------------
| top_p 1.0, sequential | 200	| 0 | 0 |
| top_p 0.95, sequential | 200 | 0 | 0 |
| greedy, sequential | 100 | 0 | 0 |
| top_p 1.0, 32-concurrent | 200 | 0 | 0 |
| top_p 0.95, 32-concurrent | 200 | 0 | 0 |
| greedy, 32-concurrent | 128 | 0 | 0 |

If we take fail rate from before of 4.5 to 9% clean run of this magnitude is extremely unlikely with previous bug so I would conclude that the top_p issue is fixed.

Closing the issue.
